SE-XXX Typed throws.md

```diff
@@ -1,38 +1,289 @@
+# Typed throws
 
-- Typed `throws`
-- Proposal: SE-NNNNN
-- Authors: @minuscorp
-- Review Manager: TBD
-- Status: Proposed
+* Proposal: [SE-NNNN](NNNN-filename.md)
+* Authors: [Jorge (@minuscorp)](https://github.com/minuscorp), [Torsten Lehmann](https://github.com/torstenlehmann)
+* Review Manager: TBD
+* Status: **Proposed**
 
-# Introduction
+## Introduction
 
-Swift Error handling system seems lacking of the other features that the Swift language has: a type system. Having typed thrown errors allow users to make a more faine-grained error handling and a more safety when facing error handling procedures.
+`throws` in Swift is missing the possibility to use it with specific error types. On the contrary [`Result`](https://developer.apple.com/documentation/swift/result) and [`Future`](https://developer.apple.com/documentation/combine/future) support specific error types. This is inconsistent without reason. The proposal is about introducing the possibility to support specific error types with `throws`.
 
+Swift-evolution thread: [Typed throw functions - Evolution / Discussion - Swift Forums](https://forums.swift.org/t/typed-throw-functions/38860)
 
-# Motivation
+## Motivation
 
-Many Swift APIs provide methods marked as `throws`, this, however, provides little to no information to the consumer about what kind of error is being thrown, if any, unless the provider adds documentation about it, which means nothing to the compiler neither the safety of the code being produced. Documentation tends to get stale and then, chances of producing unexpected errors for the user are higher. We illustrate this kind of current behaviour in the following way:
+Swift is about being explicit about semantics and using types to communicate constraints that apply to specific structures and APIs. Some developers are not satisfied with the current state of `throws` as it is not explicit about errors that are thrown. These leads to the following issues with `throws` current behaviour.
+
+### Communicates less information than `Result` or `Future`
+
+Assume you have this Error type
 
 ```swift
-import ExternalLibrary
+enum CatError: Error {
+    case sleeps
+    case sitsAtATree
+}
+```
+
+Compare
+
+```swift
+func callCat() -> Result<Cat, CatError>
+```
+
+or
+
+```swift
+func callFutureCat() -> Future<Cat, CatError>
+```
+
+with
+
+```swift
+func callCatOrThrow() throws -> Cat
+```
+
+`throws` communicates less information about why the cat is not about to come to you.
+
+### Inconsistent explicitness compared to `Result` or `Future`
+
+`throws` it's not consistent in the order of explicitness in comparison to `Result` or `Future`, which makes it hard to convert between these types or compose them easily.
+
+```swift
+func callAndFeedCat1() -> Result<Cat, CatError> {
+    do {
+        return Result.success(try callCatOrThrow())

```

Comment:
This is incorrect the correct API to convert a throwing function to a Result is [`Result(catching: )`](https://developer.apple.com/documentation/swift/result/3139399-init) but won't compile because the throwing type is `Error` and not `CatError`

Reply:
I wrote the implementation of `Result(catching: )` as far as I can see. Maybe I need another example to make the point clear. 🤔

Reply:
func throwing() throws -> String { return "Hello" }
let result: Result<String, Error> = Result<String, Error>.success(try throwing())

You're right, it takes a throwing function, my bad sorry.